### PR TITLE
plan, executor: support almost full features about generated column.

### PR DESCRIPTION
**Pull request by hicqu**
Now we can calculate generated columns in CURD.

blocked by #3754

### CODE REVIEW

File: plan/predicate_push_down.go [hunk 1]
```diff
@@ -32,6 +32,32 @@ func addSelection(p Plan, child LogicalPlan, conditions []expression.Expression,
 	return InsertPlan(p, child, selection)
 }
 
+// replaceColumnByGenerationExpression replaces all generated columns in expr by
+// their generation expressions. Returns a new Expression, maybe not changed.
+// It can be called only in DataSource.PredicatePushDown.
+func (p *DataSource) replaceColumnByGenerationExpression(expr expression.Expression) expression.Expression {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: Should we consider the `GeneratedStored` ?

File: plan/logical_plans.go [hunk 2]
```diff
@@ -328,6 +332,11 @@ type Update struct {
 	basePhysicalPlan
 
 	OrderedList []*expression.Assignment
+
+	// NormalAssignLength is for those assignments calculated by generation expressions.
+	// They are appended into OrderedList, so we need a length to know now many normal
```
Comment (tiancaiamao) [inline plan/logical_plans.go]: s/now/how

File: plan/planbuilder.go [hunk 3]
```diff
@@ -805,6 +808,70 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
 		}
 		addChild(insertPlan, selectPlan)
 	}
+
+	// Calculate generated columns if (1) it's stored or (2) there are indices on that.
+	for _, column := range insertPlan.Table.Cols() {
```
Comment (tiancaiamao) [inline plan/planbuilder.go]: And this can be another function?

File: plan/column_pruning.go [hunk 4]
```diff
@@ -157,6 +157,23 @@ func (p *Union) PruneColumns(parentUsedCols []*expression.Column) {
 // PruneColumns implements LogicalPlan interface.
 func (p *DataSource) PruneColumns(parentUsedCols []*expression.Column) {
 	used := getUsedList(parentUsedCols, p.schema)
+
+	// Resolve all generated columns.
+	var name2Offset = make(map[string]int, len(p.schema.Columns))
```
Comment (hicqu) [inline plan/column_pruning.go]: 1) In other logical plans, we don't need to resolve generated columns, so I prefer keep this logic here;
2) yes, we only needs resolve virtual generated columns.

File: executor/write.go [hunk 5]
```diff
@@ -819,15 +844,32 @@ func (e *InsertValues) getRows(cols []*table.Column) (rows [][]types.Datum, err
...
+	}
+	datums, err := e.fillRowData(cols, vals, false)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// For generated column, we should eval expr with a datum list
+	// which has same schema with e.Table. And, we can eval them one
+	// by one, because they can only refer generated columns occurring
+	// earilier in the table.
+	for i, expr := range list[length:] {
```
Comment (hicqu) [inline executor/write.go]: In `executor/write.go`, all generated columns are always coms last. If you update your table **t** like this:
```sql
update t set a = 3, b = 2, e = 6;
```
the set list will be `a = 3, b = 2, e = 6, d = 3+2`.

File: plan/logical_plans.go [hunk 6]
```diff
@@ -261,6 +261,10 @@ type DataSource struct {
 	pushedDownConds []expression.Expression
 
 	statisticTable *statistics.Table
+
+	// GenValues is for virtual generated columns without indices.
+	// the key is the offset in row datums, like expression.Column.Index.
+	GenValues map[int]expression.Expression
```
Comment (hicqu) [inline plan/logical_plans.go]: because we will use the field when build executor.

File: executor/write.go [hunk 7]
```diff
@@ -939,6 +981,9 @@ func (e *InsertValues) initDefaultValues(row []types.Datum, marked map[int]struc
 			if !e.ctx.GetSessionVars().RetryInfo.Retrying {
 				e.ctx.GetSessionVars().RetryInfo.AddAutoIncrementID(recordID)
 			}
+		} else if len(c.GeneratedExprString) != 0 {
+			// just leave generated column as null.
+			row[i].SetNull()
```
Comment (hicqu) [inline executor/write.go]: it's a bug, I'll fix it.

File: executor/distsql.go [hunk 8]
```diff
@@ -54,6 +54,21 @@ func resultRowToRow(t table.Table, h int64, data []types.Datum, tableAsName *mod
...
 
+// calculateGeneratedColumns calculates all virtual generated columns, with such arguments:
+//  cols is the collection of columns of vs;
+//  gvs's keys are index in cols, and the values are generation expressions.
+func calculateGeneratedColumns(cols []*model.ColumnInfo, gvs map[int]expression.Expression, vs []types.Datum) (err error) {
+	for i, col := range cols {
+		if len(col.GeneratedExprString) != 0 && !col.GeneratedStored {
+			vs[i], err = gvs[col.Offset].Eval(vs)
+			if err != nil {
+				return
```
Comment (tiancaiamao) [inline executor/distsql.go]: errors.Trace(err)

File: executor/distsql.go [hunk 9]
```diff
@@ -54,6 +54,21 @@ func resultRowToRow(t table.Table, h int64, data []types.Datum, tableAsName *mod
 	return &Row{Data: data, RowKeys: []*RowKeyEntry{entry}}
 }
 
+// calculateGeneratedColumns calculates all virtual generated columns, with such arguments:
+//  cols is the collection of columns of vs;
+//  gvs's keys are index in cols, and the values are generation expressions.
+func calculateGeneratedColumns(cols []*model.ColumnInfo, gvs map[int]expression.Expression, vs []types.Datum) (err error) {
+	for i, col := range cols {
+		if len(col.GeneratedExprString) != 0 && !col.GeneratedStored {
+			vs[i], err = gvs[col.Offset].Eval(vs)
```
Comment (tiancaiamao) [inline executor/distsql.go]: Would `gvs[col.Offset]` be null pointer?

File: executor/distsql.go [hunk 10]
```diff
@@ -973,6 +991,10 @@ func (e *XSelectTableExec) Next() (*Row, error) {
 			// compose aggregate row
 			return &Row{Data: values}, nil
 		}
+		// Calculate generated columns here.
+		if calculateGeneratedColumns(e.Columns, e.genValues, values) != nil {
```
Comment (tiancaiamao) [inline executor/distsql.go]:
```
if err := calculateGeneratedColumns(e.Columns, e.genValues, values); err != nil
```

File: executor/new_distsql.go [hunk 11]
```diff
@@ -50,12 +50,17 @@ type TableReaderExecutor struct {
...
-	// columns are only required by union scan.
-	columns []*model.ColumnInfo
+	columns   []*model.ColumnInfo
 
 	// result returns one or more distsql.PartialResult and each PartialResult is returned by one region.
 	result        distsql.SelectResult
 	partialResult distsql.PartialResult
+
+	// aggregate indicates whether it contains Aggregation executors or not.
+	aggregate bool
```
Comment (tiancaiamao) [inline executor/new_distsql.go]: I wonder that contains Aggregation executors affect the behavior of generated columns?

File: executor/new_distsql.go [hunk 12]
```diff
@@ -106,6 +111,11 @@ func (e *TableReaderExecutor) Next() (*Row, error) {
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
+		if !e.aggregate {
+			if calculateGeneratedColumns(e.columns, e.genValues, values) != nil {
```
Comment (tiancaiamao) [inline executor/new_distsql.go]: ditto

File: plan/logical_plan_builder.go [hunk 13]
```diff
@@ -1307,19 +1325,45 @@ func (b *planBuilder) buildUpdateLists(tableList []*ast.TableName, list []*ast.A
...
+					},
+					Expr: table.Cols()[i].GeneratedExpr,
+				})
+			}
 		}
 	}
 
 	newList := make([]*expression.Assignment, 0, p.Schema().Len())
-	for _, assign := range list {
+	for i, assign := range append(list, virtualAssignments...) {
```
Comment (tiancaiamao) [inline plan/logical_plan_builder.go]: Don't do calculation in the range expression, make new template variable.

File: plan/planbuilder.go [hunk 14]
```diff
@@ -759,6 +759,8 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
 
 	mockTablePlan := TableDual{}.init(b.allocator, b.ctx)
 	mockTablePlan.SetSchema(schema)
+
+	onDupColNames := make([]string, 0)
```
Comment (tiancaiamao) [inline plan/planbuilder.go]: `make([]string, 0, len(insert.OnDuplicate))`

File: plan/planbuilder.go (hunk 3 above)
Comment (tiancaiamao) [inline plan/planbuilder.go]: Don't think `buildInsert` is too long?

Comment (hicqu) [inline plan/planbuilder.go]: yes, I am doing this.

File: executor/distsql.go (hunk 9 above)
Comment (hicqu) [inline executor/distsql.go]: no, all generation expressions will be stored in `enValues` of `XSelectTableExec`.

File: executor/new_distsql.go (hunk 11 above)
Comment (hicqu) [inline executor/new_distsql.go]: If the aggregation is pushed down to storage engine, it will affect the behavior or generated columns.

**Comment by tiancaiamao** [comment]
LGTM
This is a big PR, please take a close look. @hanfei1991 @XuHuaiyu

File: plan/plans.go [hunk 15]
```diff
@@ -108,6 +108,14 @@ type Simple struct {
 	Statement ast.StmtNode
 }
 
+// InsertGeneratedColumns is for completing generated columns in Insert.
+// We resolve generation expressions in plan, and eval those in executor.
+type InsertGeneratedColumns struct {
+	Columns []*ast.ColumnName
```
Comment (hanfei1991) [inline plan/plans.go]: Why do we use ast.ColumnName instead of expression.Column ?

File: plan/planbuilder.go [hunk 16]
```diff
@@ -626,6 +626,44 @@ func (b *planBuilder) findDefaultValue(cols []*table.Column, name *ast.ColumnNam
...
 }
 
+func (b *planBuilder) calculateGeneratedColumns(columns []*table.Column, onDups map[string]struct{}, mockPlan LogicalPlan) (igc InsertGeneratedColumns) {
+	for _, column := range columns {
+		if len(column.GeneratedExprString) == 0 {
+			continue
+		}
+		columnName := &ast.ColumnName{Name: column.Name}
+		columnName.SetText(column.Name.O)
+		expr, _, err := b.rewrite(column.GeneratedExpr, mockPlan, nil, true)
```
Comment (hanfei1991) [inline plan/planbuilder.go]: It's unefficient to rewrite expression every time, because for the same table, the result of rewriting is also the same. Is there any way to store the expression directly ?

File: plan/logical_plan_builder.go [hunk 17]
```diff
@@ -1332,6 +1380,21 @@ func (b *planBuilder) buildUpdateLists(tableList []*ast.TableName, list []*ast.A
 	return newList, p
 }
 
+// extractTableAsNameForUpdate extracts tables' name alias for update.
+func extractTableAsNameForUpdate(p Plan) (asNames map[*model.TableInfo]model.CIStr) {
```
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: For SQL `update a from b a, (select * from a c) b`, will the tableInfo a match with name c ?

File: plan/plans.go (hunk 15 above)
Comment (hicqu) [inline plan/plans.go]: because there is `ast.ColumnName` in `Insert`. And, the offset of a column to schema is calculated in `executor`, not in `ResolveIndices`. I prefer keep it as original in this PR. If doing that in `plan` is better than `executor`, I can do that in next PR.

File: executor/distsql.go [hunk 18]
```diff
@@ -973,6 +992,10 @@ func (e *XSelectTableExec) Next() (*Row, error) {
 			// compose aggregate row
 			return &Row{Data: values}, nil
 		}
+		// Calculate generated columns here.
+		if err := calculateGeneratedColumns(e.Columns, e.genValues, values); err != nil {
```
Comment (hicqu) [inline executor/distsql.go]: I thought double reader also retrieves data from `XSelectTableExec`, right?

File: plan/planbuilder.go (hunk 16 above)
Comment (hicqu) [inline plan/planbuilder.go]: I don't think we can do that optimization, because in **column prune** phase, some columns can be droped so that the schemas from `table` and `DataSource` will be not same.

File: plan/logical_plan_builder.go [hunk 19]
```diff
@@ -1130,10 +1131,24 @@ func (b *planBuilder) buildDataSource(tn *ast.TableName) LogicalPlan {
...
+		idx := col.Position
+		colInfo := tableInfo.Columns[idx]
+		if len(colInfo.GeneratedExprString) != 0 && !colInfo.GeneratedStored {
+			column := tbl.Cols()[idx]
+			expr, _, err := b.rewrite(column.GeneratedExpr, p, nil, true)
+			if err != nil {
+				b.err = errors.Trace(err)
+				return nil
+			}
+			p.GenValues[idx] = expr
```
Comment (hicqu) [inline plan/logical_plan_builder.go]: good point, I'll try it.

File: executor/distsql.go (hunk 18 above)
Comment (hanfei1991) [inline executor/distsql.go]: Double reader reads data by itself.

Comment (hicqu) [inline executor/distsql.go]: fixed.

File: plan/logical_plan_builder.go (hunk 19 above)
Comment (hicqu) [inline plan/logical_plan_builder.go]: After try, I think do virtual column generation in `column prune` is not a better idea, because all logical plans above `DataSource` need to add some code. That's too wordy.

**Comment by shenli** [comment]
@zimulala @tiancaiamao @hanfei1991 PTAL
